<!DOCTYPE HTML>
<html lang="en" class="sidebar-visible no-js light">
    <head>
        <!-- Book generated using mdBook -->
        <meta charset="UTF-8">
        <title>PreQC Pack: A quality control tool for high throughput sequence data.</title>
        <meta name="robots" content="noindex" />


        <!-- Custom HTML head -->
        
        <meta name="description" content="">
        <meta name="viewport" content="width=device-width, initial-scale=1">
        <meta name="theme-color" content="#ffffff" />

        <link rel="icon" href="favicon.svg">
        <link rel="shortcut icon" href="favicon.png">
        <link rel="stylesheet" href="css/variables.css">
        <link rel="stylesheet" href="css/general.css">
        <link rel="stylesheet" href="css/chrome.css">
        <link rel="stylesheet" href="css/print.css" media="print">

        <!-- Fonts -->
        <link rel="stylesheet" href="FontAwesome/css/font-awesome.css">
        <link rel="stylesheet" href="fonts/fonts.css">

        <!-- Highlight.js Stylesheets -->
        <link rel="stylesheet" href="highlight.css">
        <link rel="stylesheet" href="tomorrow-night.css">
        <link rel="stylesheet" href="ayu-highlight.css">

        <!-- Custom theme stylesheets -->

    </head>
    <body>
    <div id="body-container">
        <!-- Provide site root to javascript -->
        <script>
            var path_to_root = "";
            var default_theme = window.matchMedia("(prefers-color-scheme: dark)").matches ? "navy" : "light";
        </script>

        <!-- Work around some values being stored in localStorage wrapped in quotes -->
        <script>
            try {
                var theme = localStorage.getItem('mdbook-theme');
                var sidebar = localStorage.getItem('mdbook-sidebar');

                if (theme.startsWith('"') && theme.endsWith('"')) {
                    localStorage.setItem('mdbook-theme', theme.slice(1, theme.length - 1));
                }

                if (sidebar.startsWith('"') && sidebar.endsWith('"')) {
                    localStorage.setItem('mdbook-sidebar', sidebar.slice(1, sidebar.length - 1));
                }
            } catch (e) { }
        </script>

        <!-- Set the theme before any content is loaded, prevents flash -->
        <script>
            var theme;
            try { theme = localStorage.getItem('mdbook-theme'); } catch(e) { }
            if (theme === null || theme === undefined) { theme = default_theme; }
            var html = document.querySelector('html');
            html.classList.remove('no-js')
            html.classList.remove('light')
            html.classList.add(theme);
            html.classList.add('js');
        </script>

        <!-- Hide / unhide sidebar before it is displayed -->
        <script>
            var html = document.querySelector('html');
            var sidebar = null;
            if (document.body.clientWidth >= 1080) {
                try { sidebar = localStorage.getItem('mdbook-sidebar'); } catch(e) { }
                sidebar = sidebar || 'visible';
            } else {
                sidebar = 'hidden';
            }
            html.classList.remove('sidebar-visible');
            html.classList.add("sidebar-" + sidebar);
        </script>

        <nav id="sidebar" class="sidebar" aria-label="Table of contents">
            <div class="sidebar-scrollbox">
                <ol class="chapter"><li class="chapter-item expanded "><a href="intro.html"><strong aria-hidden="true">1.</strong> Introduction</a></li><li class="chapter-item expanded "><a href="checksum.html"><strong aria-hidden="true">2.</strong> Checksum</a></li><li class="chapter-item expanded "><a href="fastqc.html"><strong aria-hidden="true">3.</strong> FastQC</a></li><li><ol class="section"><li class="chapter-item expanded "><a href="fastqc/metrics/basicStats.html"><strong aria-hidden="true">3.1.</strong> Basic Statistics</a></li><li class="chapter-item expanded "><a href="fastqc/metrics/perBaseSeqQuality.html"><strong aria-hidden="true">3.2.</strong> Per Base Sequence Quality</a></li><li class="chapter-item expanded "><a href="fastqc/metrics/perSeqQualityScore.html"><strong aria-hidden="true">3.3.</strong> Per Sequence Quality Score</a></li><li class="chapter-item expanded "><a href="fastqc/metrics/perBaseSeqContent.html"><strong aria-hidden="true">3.4.</strong> Per Base Sequence Content</a></li><li class="chapter-item expanded "><a href="fastqc/metrics/perSeqGCContent.html"><strong aria-hidden="true">3.5.</strong> Per Sequence GC Content</a></li><li class="chapter-item expanded "><a href="fastqc/metrics/perBaseNContent.html"><strong aria-hidden="true">3.6.</strong> Per Base N Content</a></li><li class="chapter-item expanded "><a href="fastqc/metrics/seqLenDistribution.html"><strong aria-hidden="true">3.7.</strong> Sequence Length Distribution</a></li><li class="chapter-item expanded "><a href="fastqc/metrics/overRepresentedSeqs.html"><strong aria-hidden="true">3.8.</strong> Overrepresented Sequences</a></li><li class="chapter-item expanded "><a href="fastqc/metrics/seqDuplicationLevel.html"><strong aria-hidden="true">3.9.</strong> Duplicate Sequences</a></li><li class="chapter-item expanded "><a href="fastqc/metrics/adapterContent.html"><strong aria-hidden="true">3.10.</strong> Adapter Content</a></li><li class="chapter-item expanded "><a href="fastqc/metrics/kmerContent.html"><strong aria-hidden="true">3.11.</strong> Kmer Content</a></li><li class="chapter-item expanded "><a href="fastqc/metrics/perTileQualityScore.html"><strong aria-hidden="true">3.12.</strong> Per Tile Sequence Quality</a></li></ol></li><li class="chapter-item expanded "><a href="mislabeling.html"><strong aria-hidden="true">4.</strong> Mislabeling</a></li></ol>
            </div>
            <div id="sidebar-resize-handle" class="sidebar-resize-handle"></div>
        </nav>

        <div id="page-wrapper" class="page-wrapper">

            <div class="page">
                                <div id="menu-bar-hover-placeholder"></div>
                <div id="menu-bar" class="menu-bar sticky bordered">
                    <div class="left-buttons">
                        <button id="sidebar-toggle" class="icon-button" type="button" title="Toggle Table of Contents" aria-label="Toggle Table of Contents" aria-controls="sidebar">
                            <i class="fa fa-bars"></i>
                        </button>
                        <button id="theme-toggle" class="icon-button" type="button" title="Change theme" aria-label="Change theme" aria-haspopup="true" aria-expanded="false" aria-controls="theme-list">
                            <i class="fa fa-paint-brush"></i>
                        </button>
                        <ul id="theme-list" class="theme-popup" aria-label="Themes" role="menu">
                            <li role="none"><button role="menuitem" class="theme" id="light">Light</button></li>
                            <li role="none"><button role="menuitem" class="theme" id="rust">Rust</button></li>
                            <li role="none"><button role="menuitem" class="theme" id="coal">Coal</button></li>
                            <li role="none"><button role="menuitem" class="theme" id="navy">Navy</button></li>
                            <li role="none"><button role="menuitem" class="theme" id="ayu">Ayu</button></li>
                        </ul>
                        <button id="search-toggle" class="icon-button" type="button" title="Search. (Shortkey: s)" aria-label="Toggle Searchbar" aria-expanded="false" aria-keyshortcuts="S" aria-controls="searchbar">
                            <i class="fa fa-search"></i>
                        </button>
                    </div>

                    <h1 class="menu-title">PreQC Pack: A quality control tool for high throughput sequence data.</h1>

                    <div class="right-buttons">
                        <a href="print.html" title="Print this book" aria-label="Print this book">
                            <i id="print-button" class="fa fa-print"></i>
                        </a>

                    </div>
                </div>

                <div id="search-wrapper" class="hidden">
                    <form id="searchbar-outer" class="searchbar-outer">
                        <input type="search" id="searchbar" name="searchbar" placeholder="Search this book ..." aria-controls="searchresults-outer" aria-describedby="searchresults-header">
                    </form>
                    <div id="searchresults-outer" class="searchresults-outer hidden">
                        <div id="searchresults-header" class="searchresults-header"></div>
                        <ul id="searchresults">
                        </ul>
                    </div>
                </div>

                <!-- Apply ARIA attributes after the sidebar and the sidebar toggle button are added to the DOM -->
                <script>
                    document.getElementById('sidebar-toggle').setAttribute('aria-expanded', sidebar === 'visible');
                    document.getElementById('sidebar').setAttribute('aria-hidden', sidebar !== 'visible');
                    Array.from(document.querySelectorAll('#sidebar a')).forEach(function(link) {
                        link.setAttribute('tabIndex', sidebar === 'visible' ? 0 : -1);
                    });
                </script>

                <div id="content" class="content">
                    <main>
                        <div style="break-before: page; page-break-before: always;"></div><h1 id="checksum"><a class="header" href="#checksum">Checksum</a></h1>
<div style="break-before: page; page-break-before: always;"></div><h1 id="fastqc"><a class="header" href="#fastqc"><a href="https://www.bioinformatics.babraham.ac.uk/projects/fastqc/Help/1%20Introduction/1.1%20What%20is%20FastQC.html">FastQC</a></a></h1>
<p>Modern high throughput sequencers can generate hundreds of millions of sequences in a single run. Before analysing this sequence to draw biological conclusions you should always perform some simple quality control checks to ensure that the raw data looks good and there are no problems or biases in your data which may affect how you can usefully use it.</p>
<p>Most sequencers will generate a QC report as part of their analysis pipeline, but this is usually only focused on identifying problems which were generated by the sequencer itself. FastQC aims to provide a QC report which can spot problems which originate either in the sequencer or in the starting library material.</p>
<p>FastQC can be run in one of two modes. It can either run as a stand alone interactive application for the immediate analysis of small numbers of FastQ files, or it can be run in a non-interactive mode where it would be suitable for integrating into a larger analysis pipeline for the systematic processing of large numbers of files.</p>
<div style="break-before: page; page-break-before: always;"></div><h1 id="basic-statistics"><a class="header" href="#basic-statistics"><a href="https://www.bioinformatics.babraham.ac.uk/projects/fastqc/Help/3%20Analysis%20Modules/1%20Basic%20Statistics.html">Basic Statistics</a></a></h1>
<h2 id="summary"><a class="header" href="#summary">Summary</a></h2>
<p>The Basic Statistics module generates some simple composition statistics for the file analysed.</p>
<ul>
<li><strong>File Name</strong>：The original filename of the file which was analysed</li>
<li><strong>File type</strong>：Says whether the file appeared to contain actual base calls or colorspace data which had to be converted to base calls</li>
<li><strong>Phred Encoding</strong>：Says which ASCII encoding of quality values was found in this file.</li>
<li><strong>Total Reads</strong>：A count of the total number of sequences processed. </li>
<li><strong>Total Bases</strong>：A count of the total number of bases in all sequences processed. </li>
<li><strong>Total T Bases</strong>：A count of the total number of base 'T' in all sequences processed. </li>
<li><strong>Total C Bases</strong>：A count of the total number of base 'C' in all sequences processed. </li>
<li><strong>Total G Bases</strong>：A count of the total number of base 'G' in all sequences processed. </li>
<li><strong>Total A Bases</strong>：A count of the total number of base 'A' in all sequences processed. </li>
<li><strong>Total N Bases</strong>： A count of the total number of base 'N' in all sequences processed , where 'N' is the base that cannot be recognized by sequencing. </li>
<li><strong>%GC</strong>：The overall %GC of all bases in all sequences</li>
<li><strong>Min Length</strong>：The length of the shortest sequence in the set.</li>
<li><strong>Max Length</strong>：The length of the highest sequence in the set.</li>
<li><strong>Lowest Char:</strong> The lowest quality char in all sequences processed. </li>
<li><strong>Highest Char:</strong> The highest quality char in all sequences processed. </li>
</ul>
<h2 id="example"><a class="header" href="#example">Example</a></h2>
<pre><code>&quot;basic_stats&quot;：{
            &quot;file_name&quot;：&quot;test.fastq.gz&quot;,
            &quot;file_type&quot;：&quot;&quot;,
            &quot;phred&quot;：{
                &quot;name&quot;：&quot;Sanger / Illumina 1.9&quot;,
                &quot;offset&quot;：33
            },
            &quot;total_reads&quot;：250000,
            &quot;total_bases&quot;：37500000,
            &quot;t_count&quot;：8473918,
            &quot;c_count&quot;：10050719,
            &quot;g_count&quot;：9888266,
            &quot;a_count&quot;：9086549,
            &quot;n_count&quot;：548,
            &quot;gc_percentage&quot;：0.5317062666666666,
            &quot;min_length&quot;：150,
            &quot;max_length&quot;：150,
            &quot;lowest_char&quot;：35,
            &quot;highest_char&quot;：70
        },
</code></pre>
<div style="break-before: page; page-break-before: always;"></div><h1 id="per-base-sequence-quality"><a class="header" href="#per-base-sequence-quality"><a href="https://www.bioinformatics.babraham.ac.uk/projects/fastqc/Help/3%20Analysis%20Modules/2%20Per%20Base%20Sequence%20Quality.html">Per Base Sequence Quality</a></a></h1>
<h2 id="summary-1"><a class="header" href="#summary-1">Summary</a></h2>
<p>The Per Base Sequence Quality module  shows an overview of the range of quality values across all bases at each position in the FastQ file.
It divides all positions into several groups  which contain a certain digit of position or the range of positions. </p>
<ul>
<li><strong>X-labels</strong>：It divides all positions into several groups  which contain a certain digit of position or the range of positions. </li>
</ul>
<ul>
<li><strong>Mean</strong>：For each position or range of position in x-labels, It calculates the average quality scores  for all sequences.</li>
<li><strong>Median</strong>：For each position or range of position in x-labels, It calculates the median quality scores  for all sequences.</li>
<li><strong>Lower Quartile</strong>：For each position or range of position in x-labels, It calculates the  lower quartile（25%） quality scores  for all sequences.</li>
<li><strong>Upper Quartile</strong>：For each position or range of position in x-labels, It calculates the higher quartile（75%） quality scores  for all sequences.</li>
<li><strong>Lowest</strong>：For each position or range of position in x-labels, It calculates the lowest quality scores  for all sequences.</li>
<li><strong>Highest</strong>：For each position or range of position in x-labels, It calculates the highest quality scores  for all sequences.</li>
</ul>
<h2 id="example-1"><a class="header" href="#example-1">Example</a></h2>
<pre><code>&quot;per_base_seq_quality&quot;: {
            &quot;xlabels&quot;: [&quot;1&quot;,&quot;2&quot;,&quot;3&quot;,&quot;4&quot;,&quot;5&quot;,&quot;6&quot;,&quot;7&quot;,&quot;8&quot;,&quot;9&quot;,&quot;10-14&quot;,&quot;15-19&quot;,&quot;20-24&quot;,&quot;25-29&quot;,&quot;30-34&quot;,&quot;35-39&quot;,...],
            &quot;mean&quot;: [36.307356,36.290732,36.46844,36.51792,36.521944,36.526816,36.473336,
36.543864,36.515792,36.563705600000006,36.560803199999995,36.5081312,36.4498688,36.42268,36.3846096, ...],
            &quot;median&quot;: [37,37,37,37,37,37,37,37,37,37,37,37,37,37,37,...],
            &quot;lower_quartile&quot;: [37,37,37,37,37,37,37,37,37,37,37,37,37,37,37,... ],
            &quot;upper_quartile&quot;: [37,37,37,37,37,37,37,37,37,37,37,37,37,37,37,... ],
            &quot;lowest&quot;: [37,37,37,37,37,37,37,37,37,37,37,37,37,37,37,... ],
            &quot;highest&quot;: [37,37,37,37,37,37,37,37,37,37,37,37,37,37,37,... ],
        },
</code></pre>
<div style="break-before: page; page-break-before: always;"></div><h1 id="per-sequence-quality-score"><a class="header" href="#per-sequence-quality-score"><a href="https://www.bioinformatics.babraham.ac.uk/projects/fastqc/Help/3%20Analysis%20Modules/3%20Per%20Sequence%20Quality%20Scores.html">Per Sequence Quality Score</a></a></h1>
<h2 id="summary-2"><a class="header" href="#summary-2">Summary</a></h2>
<p>The Per Sequence Quality Score report describes  allows you to see if a subset of your sequences have universally low quality values. It is often the case that a subset of sequences will have universally poor quality, often because they are poorly imaged (on the edge of the field of view etc), however these should represent only a small percentage of the total sequences.</p>
<p>If a significant proportion of the sequences in a run have overall low quality then this could indicate some kind of systematic problem - possibly with just part of the run (for example one end of a flowcell).</p>
<p>Results from this module will not be displayed if your input is a BAM/SAM file in which quality scores have not been recorded.</p>
<ul>
<li><strong>X- category quality</strong>：It's a set of distinct quality scores in all sequences processed, and all scores are sorted in an ascending order.</li>
<li><strong>Y-category count</strong>：It records the quality count in all sequences processed for each quality score in x-category quality.</li>
<li><strong>Most frequent score</strong>：The quality score that occurs the most</li>
</ul>
<h2 id="example-2"><a class="header" href="#example-2">Example</a></h2>
<pre><code>&quot;per_seq_quality_score&quot;: {
            &quot;x_category_quality&quot;: [16,17,18,19,20,21,22,23,24,25,...],
            &quot;y_category_count&quot;: [182111,5,4,9,37,61,126,342,479,485,...],
            &quot;most_frequent_score&quot;: 16
        },
</code></pre>
<div style="break-before: page; page-break-before: always;"></div><h1 id="per-base-sequence-content"><a class="header" href="#per-base-sequence-content"><a href="https://www.bioinformatics.babraham.ac.uk/projects/fastqc/Help/3%20Analysis%20Modules/4%20Per%20Base%20Sequence%20Content.html"> Per Base Sequence Content</a></a></h1>
<h2 id="summary-3"><a class="header" href="#summary-3">Summary</a></h2>
<p>Per Base Sequence Content module shows the proportion of each base position in a file for which each of the four normal DNA bases has been called.</p>
<p>In a random library you would expect that there would be little to no difference between the different bases of a sequence run, so the lines in this plot should run parallel with each other. The relative amount of each base should reflect the overall amount of these bases in your genome, but in any case they should not be hugely imbalanced from each other.</p>
<p>It's worth noting that some types of library will always produce biased sequence composition, normally at the start of the read. Libraries produced by priming using random hexamers (including nearly all RNA-Seq libraries) and those which were fragmented using transposases inherit an intrinsic bias in the positions at which reads start. This bias does not concern an absolute sequence, but instead provides enrichement of a number of different K-mers at the 5' end of the reads. Whilst this is a true technical bias, it isn't something which can be corrected by trimming and in most cases doesn't seem to adversely affect the downstream analysis. It will however produce a warning or error in this module.</p>
<ul>
<li><strong>X-category</strong>：It divides all base positions into several groups  which contain a certain digit of position or the range of positions. </li>
<li><strong>G counts</strong>：The counts of each base position in a file for which ’G‘ base has been called.</li>
<li><strong>C counts</strong>：The counts of each base position in a file for which ’C‘ base has been called.</li>
<li><strong>A counts</strong>：The counts of each base position in a file for which ’A‘ base has been called.</li>
<li><strong>T counts</strong>：The counts of each base position in a file for which ’T‘ base has been called.</li>
<li><strong>Percentages</strong>：The proportion of each base position in a file for which each of the four normal DNA bases has been called.Its order in  storage is <code>[t_percent_array, c_percent_array, a_percent_array,  g_percent_array]</code></li>
</ul>
<h2 id="example-3"><a class="header" href="#example-3">Example</a></h2>
<pre><code>&quot;per_base_seq_content&quot;: {
            &quot;x_category&quot;: [&quot;1&quot;,&quot;2&quot;,&quot;3&quot;,&quot;4&quot;,&quot;5&quot;,&quot;6&quot;,&quot;7&quot;,&quot;8&quot;,&quot;9&quot;,&quot;10-14&quot;,&quot;15-19&quot;,&quot;20-24&quot;,&quot;25-29&quot;,...],
            &quot;g_counts&quot;: [
            	115472,
                80383,
                76362,
                83320,
                81674,
                80780,
                61621,
                59846,
                61399,
                60254,
                67033,
                64991,
                65134,
                ...
            ],
            &quot;c_counts&quot;: [
            	63849,
                69603,
                80458,
                75224,
                60497,
                51506,
                49210,
                57837,
                55266,
                54779,
                59960,
                60946,
                61423,
                ...
            ],
            &quot;a_counts&quot;: [
            	38933,
                29960,
                35063,
                41477,
                50424,
                59984,
                60756,
                62327,
                58272,
                72997,
                65287,
                64525,
                64301,
                ...
            ],
            &quot;t_counts&quot;: [
            	31663,
                69593,
                58117,
                49979,
                57405,
                57728,
                78413,
                69990,
                75063,
                61970,
                57718,
                59538,
                59142,
                ...
            ],
            &quot;percentages&quot;: [
                [
                    12.669406242872633,
                    27.88862662750111,
                    23.2468,
                    19.991600000000002,
                    22.962,
                    23.09138473107785,
                    31.365199999999998,
                    27.996,
                    30.0252,
                    23.834518135229015,
                    22.720000000000002,
                    23.19664,
                    22.7348,
                    ...
                ],
                [
                    25.54808196321179,
                    27.892634017127584,
                    32.1832,
                    30.0896,
                    24.198800000000002,
                    20.602564820518566,
                    19.683999999999997,
                    23.1348,
                    22.1064,
                    23.88315821305314,
                    25.505119999999998,
                    25.90864,
                    27.118,
                    ...
                ],
                [
                    15.578372019510478,
                    12.006139320907753,
                    14.025199999999998,
                    16.5908,
                    20.1696,
                    23.993791950335602,
                    24.3024,
                    24.9308,
                    23.308799999999998,
                    26.43276229241967,
                    24.285680000000003,
                    24.44032,
                    25.01016,
                    ...
                ],
                [
                    46.2041397744051,
                    32.21260003446355,
                    30.5448,
                    33.328,
                    32.669599999999996,
                    32.31225849806798,
                    24.648400000000002,
                    23.9384,
                    24.5596,
                    25.849561359298175,
                    27.489200000000004,
                    26.4544,
                    25.13704,
                    ...
                ]
            ]
  },
</code></pre>
<div style="break-before: page; page-break-before: always;"></div><h1 id="per-sequence-gc-content"><a class="header" href="#per-sequence-gc-content"><a href="https://www.bioinformatics.babraham.ac.uk/projects/fastqc/Help/3%20Analysis%20Modules/5%20Per%20Sequence%20GC%20Content.html"> Per Sequence GC Content</a></a></h1>
<h2 id="summary-4"><a class="header" href="#summary-4">Summary</a></h2>
<p>This module measures the GC content across the whole length of each sequence in a file and compares it to a modelled normal distribution of GC content.</p>
<p>In a normal random library you would expect to see a roughly normal distribution of GC content where the central peak corresponds to the overall GC content of the underlying genome. Since we don't know the the GC content of the genome the modal GC content is calculated from the observed data and used to build a reference distribution.</p>
<p>An unusually shaped distribution could indicate a contaminated library or some other kinds of biased subset. A normal distribution which is shifted indicates some systematic bias which is independent of base position. If there is a systematic bias which creates a shifted normal distribution then this won't be flagged as an error by the module since it doesn't know what your genome's GC content should be.</p>
<ul>
<li><strong>X-category</strong>：It's a set of mean %GC. </li>
<li><strong>Y-gc distribution</strong>：For each %GC in x-category，we count the number of corresponding sentences.</li>
<li><strong>Y-theoretic distribution</strong>：  The theoretic number of corresponding sentences for each %GC in x-category.</li>
<li><strong>Deviation percent</strong>：Records the sum of the deviations for all corresponding positions of y-gc distribution and y-theoretic distribution.</li>
</ul>
<h2 id="example-4"><a class="header" href="#example-4">Example</a></h2>
<pre><code>&quot;per_seq_gc_content&quot;: {
            &quot;x_category&quot;: [
                0,
                1,
                2,
                3,
                4,
                5,
                6,
                7,
                8,
                9,
                10,
                11,
                12,
                13,
                ...
            ],
            &quot;y_gc_distribution&quot;: [
                0,
                0,
                0.5,
                0.5,
                0,
                0,
                0,
                1,
                1,
                3.5,
                3.5,
                0,
                3,
                8,
               ...
            ],
            &quot;y_theo_distribution&quot;: [
                11.301763218607428,
                13.897288489846439,
                17.029620470668664,
                20.79557306799733,
                25.306252897428912,
                30.68851136823189,
                37.08641280899335,
                44.66268842218363,
                53.60013975983366,
                64.10294913058424,
                76.39784805677265,
                90.73508881215378,
                107.3891584351843,
               ...
            ],
            &quot;deviation_percent&quot;: 42.145681622216166
        },
</code></pre>
<div style="break-before: page; page-break-before: always;"></div><h1 id="per-base-n-content"><a class="header" href="#per-base-n-content"><a href="https://www.bioinformatics.babraham.ac.uk/projects/fastqc/Help/3%20Analysis%20Modules/6%20Per%20Base%20N%20Content.html"> Per Base N Content</a></a></h1>
<h2 id="summary-5"><a class="header" href="#summary-5">Summary</a></h2>
<p>If a sequencer is unable to make a base call with sufficient confidence then it will normally substitute an N rather than a conventional base] call.</p>
<p>It's not unusual to see a very low proportion of Ns appearing in a sequence, especially nearer the end of a sequence. However, if this proportion rises above a few percent it suggests that the analysis pipeline was unable to interpret the data well enough to make valid base calls.</p>
<ul>
<li><strong>X-category</strong>：It divides all base positions into several groups  which contains a certain digit of position or the range of positions. </li>
<li><strong>N counts</strong>：Records the count of base 'N' for each position in x-category.</li>
<li><strong>Not n counts</strong>：  Records the sum of count of base 'A', base 'T', base 'C' and base 'G' for each position in x-category.</li>
<li><strong>Percentages</strong>：The proportion of each base position in x-category for which the base 'N' has been called.</li>
</ul>
<h2 id="example-5"><a class="header" href="#example-5">Example</a></h2>
<pre><code>&quot;per_base_n_content&quot;: {
           &quot;x_categories&quot;: [
                &quot;1&quot;,
                &quot;2&quot;,
                &quot;3&quot;,
                &quot;4&quot;,
                &quot;5&quot;,
                &quot;6&quot;,
                &quot;7&quot;,
                &quot;8&quot;,
                &quot;9&quot;,
                &quot;10-14&quot;,
                &quot;15-19&quot;,
                &quot;20-24&quot;,
                ...
            ],
            &quot;n_counts&quot;: [
                83,
                461,
                0,
                0,
                0,
                2,
                0,
                0,
                0,
                0,
                2,
                0,
                ...
            ],
            &quot;not_n_counts&quot;: [
                249917,
                249539,
                250000,
                250000,
                250000,
                249998,
                250000,
                250000,
                250000,
                250000,
                249998,
                250000,
                ...
            ],
            &quot;percentages&quot;: [
                0.0332,
                0.18439999999999998,
                0,
                0,
                0,
                0.0007999999999999999,
                0,
                0,
                0,
                0.00015999999999999999,
                0,
                0,
                ...
            ] 
        },
			
</code></pre>
<div style="break-before: page; page-break-before: always;"></div><h1 id="sequence-length-distribution"><a class="header" href="#sequence-length-distribution"><a href="https://www.bioinformatics.babraham.ac.uk/projects/fastqc/Help/3%20Analysis%20Modules/7%20Sequence%20Length%20Distribution.html"> Sequence Length Distribution</a></a></h1>
<h2 id="summary-6"><a class="header" href="#summary-6">Summary</a></h2>
<p>Some high throughput sequencers generate sequence fragments of uniform length, but others can contain reads of wildly varying lengths. Even within uniform length libraries some pipelines will trim sequences to remove poor quality base calls from the end.</p>
<p>This module shows the distribution of fragment sizes in the file which was analysed.</p>
<p>In many cases this will produce a simple graph showing a peak only at one size, but for variable length FastQ files this will show the relative amounts of each different size of sequence fragment.</p>
<ul>
<li><strong>X-category</strong>：A set of length.</li>
<li><strong>Graph counts</strong>：The number of sentences for each length in x-category.</li>
</ul>
<h2 id="example-6"><a class="header" href="#example-6">Example</a></h2>
<pre><code>&quot;seq_len_distribution&quot;: {
            &quot;x_categories&quot;: [
                &quot;149&quot;,
                &quot;150&quot;,
                &quot;151&quot;
            ],
            &quot;graph_counts&quot;: [
                0,
                250000,
                0
            ]
        },
</code></pre>
<div style="break-before: page; page-break-before: always;"></div><h1 id="overrepresented-sequences"><a class="header" href="#overrepresented-sequences"><a href="https://www.bioinformatics.babraham.ac.uk/projects/fastqc/Help/3%20Analysis%20Modules/9%20Overrepresented%20Sequences.html"> Overrepresented Sequences</a></a></h1>
<h2 id="summary-7"><a class="header" href="#summary-7">Summary</a></h2>
<p>A normal high-throughput library will contain a diverse set of sequences, with no individual sequence making up a tiny fraction of the whole. Finding that a single sequence is very overrepresented in the set either means that it is highly biologically significant, or indicates that the library is contaminated, or not as diverse as you expected.</p>
<p>This module lists all of the sequence which make up more than 0.1% of the total. To conserve memory only sequences which appear in the first 100,000 sequences are tracked to the end of the file. It is therefore possible that a sequence which is overrepresented but doesn't appear at the start of the file for some reason could be missed by this module.</p>
<p>For each overrepresented sequence the program will look for matches in a database of common contaminants and will report the best hit it finds. Hits must be at least 20bp in length and have no more than 1 mismatch. Finding a hit doesn't necessarily mean that this is the source of the contamination, but may point you in the right direction. It's also worth pointing out that many adapter sequences are very similar to each other so you may get a hit reported which isn't technically correct, but which has very similar sequence to the actual match.</p>
<p>Because the duplication detection requires an exact sequence match over the whole length of the sequence any reads over 75bp in length are truncated to 50bp for the purposes of this analysis. Even so, longer reads are more likely to contain sequencing errors which will artificially increase the observed diversity and will tend to underrepresent highly duplicated sequences.</p>
<ul>
<li><strong>count</strong>：Total number of  sequences in the file.</li>
<li><strong>Obervation cut off</strong>：The number of unique sequences we want to track in overrepresented module.</li>
<li><strong>Unique sequence count</strong>：The number of unique sequences at present.</li>
<li><strong>Count at unique limit</strong>： The corresponding <code>count</code> that has been processed when <code>unique sequence count</code>  reaches the <code>obervation cut off</code>.</li>
<li><strong>Overrepresented sequences</strong> : It's a set of overrepresented  sequence with attributes &quot;seq&quot;，“count”，&quot;percentage&quot; and contaminant hit&quot;, where &quot;seq&quot; is stand for a sequence of bases， “count” is stand for  the number of sentences containing such base sequence, &quot;percentage&quot; is stand for the proportion of &quot;count&quot; in total count, and contaminant hit is stand for the hit results in file of comtaminant .</li>
</ul>
<h2 id="example-7"><a class="header" href="#example-7">Example</a></h2>
<pre><code>&quot;overrepresented_seqs&quot;: {
            &quot;count&quot;: 250000,
            &quot;observation_cut_off&quot;: 100000,
            &quot;unique_seq_count&quot;: 100000,
            &quot;count_at_unique_limit&quot;: 140053,
            &quot;overrepresented_seqs&quot;: [
                {
                    &quot;seq&quot;: &quot;GTGGCTATTCACAGGCGCGATCCCACTACTGATCAGCACGGGAGTTTTGA&quot;,
                    &quot;count&quot;: 1941,
                    &quot;percentage&quot;: 0.7764,
                    &quot;contaminant_hit&quot;: null
                },
                {
                    &quot;seq&quot;: &quot;GCAGTGGCTATTCACAGGCGCGATCCCACTACTGATCAGCACGGGAGTTT&quot;,
                    &quot;count&quot;: 1347,
                    &quot;percentage&quot;: 0.5388000000000001,
                    &quot;contaminant_hit&quot;: null
                },
                {
                    &quot;seq&quot;: &quot;AGTGGCTATTCACAGGCGCGATCCCACTACTGATCAGCACGGGAGTTTTG&quot;,
                    &quot;count&quot;: 1291,
                    &quot;percentage&quot;: 0.5164,
                    &quot;contaminant_hit&quot;: null
                },
                {
                    &quot;seq&quot;: &quot;GTGCAGTGGCTATTCACAGGCGCGATCCCACTACTGATCAGCACGGGAGT&quot;,
                    &quot;count&quot;: 916,
                    &quot;percentage&quot;: 0.3664,
                    &quot;contaminant_hit&quot;: null
                },
                ...
            ]
        },
</code></pre>
<div style="break-before: page; page-break-before: always;"></div><h1 id="duplicate-sequence"><a class="header" href="#duplicate-sequence"><a href="https://www.bioinformatics.babraham.ac.uk/projects/fastqc/Help/3%20Analysis%20Modules/8%20Duplicate%20Sequences.html"> Duplicate Sequence</a></a></h1>
<h2 id="summary-8"><a class="header" href="#summary-8">Summary</a></h2>
<p>In a diverse library most sequences will occur only once in the final set. A low level of duplication may indicate a very high level of coverage of the target sequence, but a high level of duplication is more likely to indicate some kind of enrichment bias (eg PCR over amplification).</p>
<p>This module counts the degree of duplication for every sequence in a library and shows the relative number of sequences with different degrees of duplication.</p>
<p>To cut down on the memory requirements for this module only sequences which first appear in the first 100,000 sequences in each file are analysed, but this should be enough to get a good impression for the duplication levels in the whole file. Each sequence is tracked to the end of the file to give a representative count of the overall duplication level. To cut down on the amount of information any sequences with more than 10 duplicates are placed into grouped bins to give a clear impression of the overall duplication level without having to show each individual duplication value.</p>
<p>Because the duplication detection requires an exact sequence match over the whole length of the sequence, any reads over 75bp in length are truncated to 50bp for the purposes of this analysis. Even so, longer reads are more likely to contain sequencing errors which will artificially increase the observed diversity and will tend to underrepresent highly duplicated sequences.</p>
<p>The module also calculates an expected overall loss of sequence were the library to be deduplicated.</p>
<ul>
<li><strong>Total percentages</strong>：It takes the full sequence set and shows how its duplication levels are distributed. </li>
<li><strong>Deduplicated percentages</strong>：The sequences are de-duplicated and the proportions  are the proportions of the deduplicated set which come from different duplication levels in the original data.</li>
<li><strong>Percentage diff</strong>： Calculates an expected overall loss of sequence were the library to be deduplicated.</li>
<li><strong>labels</strong> : It's a set which shows different degrees of duplication.</li>
</ul>
<h2 id="example-8"><a class="header" href="#example-8">Example</a></h2>
<pre><code>&quot;dedup_percentages&quot;: [
                88.7270625897991,
                8.234690198626954,
                1.3275601587718149,
                0.4567248391318406,
                0.23253294096467203,
                0.16711084299112353,
                0.10936917951231077,
                0.08366047968337169,
                0.060472876913210516,
                0.4965347445786828,
                0.05984315938476192,
                0.041475456999339945,
                0.0011850130571239984,
                0.0017775195856859977,
                0,
                0
            ],
            &quot;total_percentages&quot;: [
                62.5830541835812,
                11.616569913245959,
                2.809157665879907,
                1.2885915309298914,
                0.8200779570004981,
                0.7072232509214026,
                0.5399998558899416,
                0.47207498413461896,
                0.383887339423591,
                7.222186751326263,
                2.92962348084703,
                6.041878268559988,
                0.6720163419689051,
                1.9136584762908062,
                0,
                0
            ],
            &quot;percent_diff_seq&quot;: 70.53434697022905,
            &quot;labels&quot;: [
                &quot;1&quot;,
                &quot;2&quot;,
                &quot;3&quot;,
                &quot;4&quot;,
                &quot;5&quot;,
                &quot;6&quot;,
                &quot;7&quot;,
                &quot;8&quot;,
                &quot;9&quot;,
                &quot;&gt;10&quot;,
                &quot;&gt;50&quot;,
                &quot;&gt;100&quot;,
                &quot;&gt;500&quot;,
                &quot;&gt;1k&quot;,
                &quot;&gt;5k&quot;,
                &quot;&gt;10k&quot;
            ]
</code></pre>
<div style="break-before: page; page-break-before: always;"></div><h1 id="adapter-content"><a class="header" href="#adapter-content"><a href="https://www.bioinformatics.babraham.ac.uk/projects/fastqc/Help/3%20Analysis%20Modules/10%20Adapter%20Content.html"> Adapter Content</a></a></h1>
<h2 id="summary-9"><a class="header" href="#summary-9">Summary</a></h2>
<p>The Kmer Content module will do a generic analysis of all of the Kmers in your library to find those which do not have even coverage through the length of your reads. This can find a number of different sources of bias in the library which can include the presence of read-through adapter sequences building up on the end of your sequences.</p>
<p>You can however find that the presence of any overrepresented sequences in your library (such as adapter dimers) will cause the Kmer plot to be dominated by the Kmers these sequences contain, and that it's not always easy to see if there are other biases present in which you might be interested.</p>
<p>One obvious class of sequences which you might want to analyse are adapter sequences. It is useful to know if your library contains a significant amount of adapter in order to be able to assess whether you need to adapter trim or not. Although the Kmer analysis can theoretically spot this kind of contamination it isn't always clear. This module therefore does a specific search for a set of separately defined Kmers and will give you a view of the total proportion of your library which contain these Kmers. A results trace will always be generated for all of the sequences present in the adapter config file so you can see the adapter content of your library, even if it's low.</p>
<p>Once a sequence has been seen in a read it is counted as being present right through to the end of the read so the percentages you see will only increase as the read length goes on</p>
<p>The module also calculates an expected overall loss of sequence were the library to be deduplicated.</p>
<ul>
<li><strong>Labels</strong>：A set of adpaters from the file of adpater.</li>
<li><strong>X-labels</strong>：It divides all positions into several groups  which contain a certain digit of position or the range of positions. </li>
<li><strong>Enrichments</strong>： It ranges in the order of adapters in <code>Labels</code> and shows a cumulative percentage count of the proportion of your library which has seen each of the adapter sequences at each position.</li>
</ul>
<h2 id="example-9"><a class="header" href="#example-9">Example</a></h2>
<pre><code>&quot;adapter_content&quot;: {
            &quot;labels&quot;: [
                &quot;Illumina Universal Adapter&quot;,
                &quot;Illumina Small RNA 3' Adapter&quot;,
                &quot;Illumina Small RNA 5' Adapter&quot;,
                &quot;Nextera Transposase Sequence&quot;,
                &quot;SOLID Small RNA Adapter&quot;
            ],
            &quot;x_labels&quot;: [
                &quot;1&quot;,
                &quot;2&quot;,
                &quot;3&quot;,
                &quot;4&quot;,
                &quot;5&quot;,
                &quot;6&quot;,
                &quot;7&quot;,
                &quot;8&quot;,
                &quot;9&quot;,
                &quot;10-11&quot;,
                &quot;12-13&quot;,
                &quot;14-15&quot;,
                &quot;16-17&quot;,
                ...
            ],
            &quot;enrichments&quot;: [
                [
                    0.002,
                    0.002,
                    0.002,
                    0.002,
                    0.002,
                    0.002,
                    0.002,
                    0.002,
                    0.002,
                    0.0021999999999999997,
                    0.0024,
                    0.0024,
                    0.0024,
                    ...
                ],
                [
                    0,
                    0,
                    0,
                    0,
                    0,
                    0,
                    0,
                    0,
                    0,
                    0,
                    0,
                    0,
                    0,
                    ...
                ],
                [
                    0,
                    0,
                    0,
                    0,
                    0,
                    0,
                    0,
                    0,
                    0,
                    0,
                    0,
                    0,
                    0,
                    ...
                ],
                [
                    0,
                    0,
                    0,
                    0,
                    0,
                    0,
                    0,
                    0,
                    0,
                    0,
                    0,
                    0,
                    0,
                    ...
                ],
                [
                    0,
                    0,
                    0,
                    0,
                    0,
                    0,
                    0,
                    0,
                    0,
                    0,
                    0,
                    0,
                    0,
                    ...
                ]
            ]
        },
</code></pre>
<div style="break-before: page; page-break-before: always;"></div><h1 id="kmer-content"><a class="header" href="#kmer-content"><a href="https://www.bioinformatics.babraham.ac.uk/projects/fastqc/Help/3%20Analysis%20Modules/10%20Adapter%20Content.html">Kmer Content</a></a></h1>
<h2 id="summary-10"><a class="header" href="#summary-10">Summary</a></h2>
<p>The analysis of overrepresented sequences will spot an increase in any exactly duplicated sequences, but there are a different subset of problems where it will not work.</p>
<ul>
<li>If you have very long sequences with poor sequence quality then random sequencing errors will dramatically reduce the counts for exactly duplicated sequences.</li>
<li>If you have a partial sequence which is appearing at a variety of places within your sequence then this won't be seen either by the per base content module or the duplicate sequence analysis.</li>
</ul>
<p>The Kmer module starts from the assumption that any small fragment of sequence should not have a positional bias in its apearance within a diverse library. There may be biological reasons why certain Kmers are enriched or depleted overall, but these biases should affect all positions within a sequence equally. This module therefore measures the number of each 7-mer at each position in your library and then uses a binomial test to look for significant deviations from an even coverage at all positions. Any Kmers with positionally biased enrichment are reported. The top 6 most biased Kmer are additionally reported to show their distribution.</p>
<p>To allow this module to run in a reasonable time only 2% of the whole library is analysed and the results are extrapolated to the rest of the library. Sequences longer than 500bp are truncated to 500bp for this analysis.</p>
<ul>
<li><strong>Enriched Kmers</strong>： This is the full set of Kmers to be reported.</li>
<li><strong>Enrichments</strong>：Significant deviations from an even coverage at all positions for each kmer in x-labels.</li>
<li><strong>X-categories</strong>：It divides all positions into several groups  which contain a certain digit of position or the range of positions. </li>
<li><strong>X-labels</strong>：A set of the top 6 most biased Kmer are additionally plotted to show their distribution.</li>
</ul>
<h2 id="example-10"><a class="header" href="#example-10">Example</a></h2>
<pre><code>        &quot;kmer_content&quot;: {
            &quot;enriched_kmers&quot;: [
                {
                    &quot;sequence&quot;: &quot;CGCATTT&quot;,
                    &quot;count&quot;: 7,
                    &quot;lowest_pvalue&quot;: 0.0036627573317673523
                },
                {
                    &quot;sequence&quot;: &quot;CTCGCTA&quot;,
                    &quot;count&quot;: 56,
                    &quot;lowest_pvalue&quot;: 0
                },
                {
                    &quot;sequence&quot;: &quot;CCCCTAT&quot;,
                    &quot;count&quot;: 14,
                    &quot;lowest_pvalue&quot;: 0.0010821496325661428
                },
            	...
            ],
            &quot;enrichments&quot;: [
                [
                    0,
                    0,
                    61.712828571428574,
                    0,
                    0,
                    0,
                    0,
                    ...
                ],
                [
                    43.792079314194126,
                    23.179397750686814,
                    10.28547142857143,
                    7.714103571428572,
                    7.714103571428572,
                    7.714103571428572,
                    ...
                ],
                [
                    41.2160746486533,
                    10.301954555860807,
                    0,
                    0,
                    0,
                    0,
                    ...
                ],
                [
                    0,
                    0,
                    20.57094285714286,
                    0,
                    30.856414285714287,
                    41.14188571428572,
                    ...
                ],
                [
                    18.032032658785813,
                    40.56394606370192,
                    20.249521875,
                    8.9997875,
                    6.749840625,
                    6.749840625,
                    ...
                ],
                [
                    6.272011359577675,
                    16.72201319212189,
                    39.65123768115942,
                    18.782165217391306,
                    8.347628985507248,
                    6.260721739130435,
                    ...
                ]
            ],
            &quot;x_categories&quot;: [
                &quot;1&quot;,
                &quot;2&quot;,
                &quot;3&quot;,
                &quot;4&quot;,
                &quot;5&quot;,
                &quot;6&quot;,
                &quot;7&quot;,
                &quot;8&quot;,
                &quot;9&quot;,
                &quot;10-14&quot;,
                &quot;15-19&quot;,
                &quot;20-24&quot;,
                ...
            ],
            &quot;x_labels&quot;: [
                &quot;CGCATTT&quot;,
                &quot;CTCGCTA&quot;,
                &quot;CCCCTAT&quot;,
                &quot;CGATTTG&quot;,
                &quot;TCGCTAT&quot;,
                &quot;CGCTATG&quot;
            ]
        },
</code></pre>
<div style="break-before: page; page-break-before: always;"></div><h1 id="per-tile-sequence-quality"><a class="header" href="#per-tile-sequence-quality"><a href="https://www.bioinformatics.babraham.ac.uk/projects/fastqc/Help/3%20Analysis%20Modules/12%20Per%20Tile%20Sequence%20Quality.html">Per Tile Sequence Quality</a></a></h1>
<h2 id="summary-11"><a class="header" href="#summary-11">Summary</a></h2>
<p>This module will only appear in your analysis results if you're using an Illumina library which retains its original sequence identifiers. Encoded in these is the flowcell tile from which each read came. The module allows you to look at the quality scores from each tile across all of your bases to see if there was a loss in quality associated with only one part of the flowcell.</p>
<ul>
<li>
<p><strong>X-labels</strong>：It divides all positions into several groups  which contain a certain digit of position or the range of positions. </p>
</li>
<li>
<p><strong>Tiles</strong>：A set of all tiles id found in file.</p>
</li>
<li>
<p><strong>Means</strong>：  The set of  average quality for each tile.</p>
</li>
</ul>
<h2 id="example-11"><a class="header" href="#example-11">Example</a></h2>
<pre><code>&quot;per_tile_quality_score&quot;: {
            
            &quot;x_labels&quot;: [
                &quot;1&quot;,
                &quot;2&quot;,
                &quot;3&quot;,
                &quot;4&quot;,
                &quot;5&quot;,
                &quot;6&quot;,
                &quot;7&quot;,
                &quot;8&quot;,
                ...
            ],
            &quot;tiles&quot;: [
                1101,
                1102,
                1103,
                1104,
                1105,
                1106,
                1107,
                1108,
                1109,
                1110,
                1111,
                1112
            ],
            &quot;means&quot;: [
                [
                    -0.006346050119049096,
                    -0.0654172611747228,
                    0.04078626086620574,
                    0.018562823162376674,
                    -0.07341049875898875,
                    -0.016429787792191064,
                    -0.01372768334803709,
                    -0.03346466757579947,
                    ...
                ],
                [
                    -0.0936709871750665,
                    -0.0331261565078762,
                    0.06435934332282045,
                    -0.037159459006495865,
                    0.03486157659394706,
                    0.006024915022308619,
                    0.01689568685248588,
                    -0.008014505719849296,
                    ...
                ],
                [
                    0.19160803277671334,
                    0.0720129699881511,
                    0.046999435285798086,
                    -0.016470494564551075,
                    0.08760990913324918,
                    -0.056012653372455645,
                    0.05859393421414438,
                    -0.0757104323515776,
                    ...
                ],
                [
                    -0.08292170351253958,
                    -0.04457089187183527,
                    -0.02132341694053963,
                    0.09498880685247002,
                    -0.02909020660642625,
                    -0.08919508767951356,
                    0.06544754755240234,
                    0.04402972234209557,
                    ...
                ],
                [
                    -0.05259925769379237,
                    0.17734065966617152,
                    0.05323547448775656,
                    -0.031241010441334538,
                    0.06060992018126399,
                    0.057363400444444324,
                    0.049313153043023306,
                    0.036693847332685436,
                    ...
                ],
                [
                    0.04623782481657912,
                    -0.1742257660234543,
                    0.008555611227521354,
                    0.0780235468237791,
                    0.05726359559632499,
                    0.07284318681385571,
                    -0.4294163047979822,
                    0.03223193133478475,
                    ...
                ],
                [
                    0.006707578514678403,
                    0.1224868275087374,
                    0.0819144995306047,
                    -0.0927164958223301,
                    0.07133213032459196,
                    -0.021621178716586087,
                    -0.02652805075825171,
                    0.024320665561241128,
                    ...
                ],
                [
                    -0.1892194282145212,
                    0.05937301539450601,
                    -0.009251856831049565,
                    0.00966011187865945,
                    -0.0506608728653859,
                    -0.000005505544429240672,
                    0.0768259129498503,
                    -0.021634040275564814,
                    ...
                ],
                [
                    -0.01482087894996198,
                    -0.04388643789315694,
                    -0.15656598135250732,
                    -0.02290621225324685,
                    -0.02460270912033735,
                    0.05281604056343525,
                    -0.027855395398596272,
                    -0.02307402952937565,
                    ...
                ],
                [
                    0.0997225814284306,
                    0.023947398136456854,
                    0.00972046412779548,
                    0.025144240173403887,
                    -0.015702242983785197,
                    -0.03782832466826136,
                    0.21077909870614775,
                    0.006800753397236292,
                    ...
                ],
                [
                    0.033955607721260606,
                    -0.09951870036752553,
                    -0.08152383052087941,
                    -0.07023362168592229,
                    -0.06128076790366066,
                    -0.03250179949124998,
                    0.07098260870256468,
                    0.001405350739482003,
                    ...
                ],
                [
                    0.06134668040730418,
                    0.005584343144498405,
                    -0.03690600320344117,
                    0.0443477648832129,
                    -0.05692983359072201,
                    0.06454679442065725,
                    -0.05131050771772294,
                    0.016415404744613227,
                    ...
                ]
            ]
        }
</code></pre>
<div style="break-before: page; page-break-before: always;"></div><h1 id="mislabeling"><a class="header" href="#mislabeling">Mislabeling</a></h1>

                    </main>

                    <nav class="nav-wrapper" aria-label="Page navigation">
                        <!-- Mobile navigation buttons -->


                        <div style="clear: both"></div>
                    </nav>
                </div>
            </div>

            <nav class="nav-wide-wrapper" aria-label="Page navigation">

            </nav>

        </div>




        <script>
            window.playground_copyable = true;
        </script>


        <script src="elasticlunr.min.js"></script>
        <script src="mark.min.js"></script>
        <script src="searcher.js"></script>

        <script src="clipboard.min.js"></script>
        <script src="highlight.js"></script>
        <script src="book.js"></script>

        <!-- Custom JS scripts -->

        <script>
        window.addEventListener('load', function() {
            window.setTimeout(window.print, 100);
        });
        </script>

    </div>
    </body>
</html>
